File: binding/src/libqsocketnotifier.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsocketnotifier.html
pub const qsocketnotifier = struct {

    /// New constructs a new QSocketNotifier object.
    pub fn New(param1: i32) QtC.QSocketNotifier {
        return qtc.QSocketNotifier_new(@intCast(param1));
    }


    /// New2 constructs a new QSocketNotifier object.
    pub fn New2(socket: isize, param2: i32) QtC.QSocketNotifier {
        return qtc.QSocketNotifier_new2(socket, @intCast(param2));
    }


    /// New3 constructs a new QSocketNotifier object.
    pub fn New3(param1: i32, parent: ?*anyopaque) QtC.QSocketNotifier {
        return qtc.QSocketNotifier_new3(@intCast(param1), @ptrCast(parent));
    }


    /// New4 constructs a new QSocketNotifier object.
    pub fn New4(socket: isize, param2: i32, parent: ?*anyopaque) QtC.QSocketNotifier {
        return qtc.QSocketNotifier_new4(socket, @intCast(param2), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSocketNotifier_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsocketnotifier.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetSocket(self: ?*anyopaque, socket: isize) void {
        qtc.QSocketNotifier_SetSocket(@ptrCast(self), socket);
    }

    pub fn Socket(self: ?*anyopaque, ) isize {
        return qtc.QSocketNotifier_Socket(@ptrCast(self));
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QSocketNotifier_Type(@ptrCast(self));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QSocketNotifier_IsValid(@ptrCast(self));
    }

    pub fn IsEnabled(self: ?*anyopaque, ) bool {
        return qtc.QSocketNotifier_IsEnabled(@ptrCast(self));
    }

    pub fn SetEnabled(self: ?*anyopaque, enabled: bool) void {
        qtc.QSocketNotifier_SetEnabled(@ptrCast(self), enabled);
    }

    pub fn Event(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QSocketNotifier_Event(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QSocketNotifier_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QSocketNotifier_QBaseEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSocketNotifier_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsocketnotifier.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSocketNotifier_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsocketnotifier.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Wrapper to allow calling private signal
    pub fn Onactivated(self: ?*anyopaque, callback: *const fn (?*anyopaque, QtC.QSocketDescriptor, i32) callconv(.c) void) void {
        qtc.QSocketNotifier_Connect_activated(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSocketNotifier_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qsocketdescriptor.html
pub const qsocketdescriptor = struct {

    /// New constructs a new QSocketDescriptor object.
    pub fn New(other: ?*anyopaque) QtC.QSocketDescriptor {
        return qtc.QSocketDescriptor_new(@ptrCast(other));
    }


    /// New2 constructs a new QSocketDescriptor object.
    pub fn New2(other: ?*anyopaque) QtC.QSocketDescriptor {
        return qtc.QSocketDescriptor_new2(@ptrCast(other));
    }


    /// New3 constructs a new QSocketDescriptor object.
    pub fn New3() QtC.QSocketDescriptor {
        return qtc.QSocketDescriptor_new3();
    }


    /// New4 constructs a new QSocketDescriptor object.
    pub fn New4(desc: isize) QtC.QSocketDescriptor {
        return qtc.QSocketDescriptor_new4(desc);
    }


    /// New5 constructs a new QSocketDescriptor object.
    pub fn New5(param1: ?*anyopaque) QtC.QSocketDescriptor {
        return qtc.QSocketDescriptor_new5(@ptrCast(param1));
    }


    /// New6 constructs a new QSocketDescriptor object.
    pub fn New6(descriptor: ?*anyopaque) QtC.QSocketDescriptor {
        return qtc.QSocketDescriptor_new6(@ptrCast(descriptor));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSocketDescriptor_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSocketDescriptor_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn OperatorlongLong(self: ?*anyopaque, ) isize {
        return qtc.QSocketDescriptor_OperatorlongLong(@ptrCast(self));
    }

    pub fn WinHandle(self: ?*anyopaque, ) ?*anyopaque {
        return qtc.QSocketDescriptor_WinHandle(@ptrCast(self));
    }

    pub fn OperatorvoidMultiply(self: ?*anyopaque, ) ?*anyopaque {
        return qtc.QSocketDescriptor_OperatorvoidMultiply(@ptrCast(self));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QSocketDescriptor_IsValid(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSocketDescriptor_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/socketnotifier.html#types
pub const enums = struct {
    pub const Type = enum {
        pub const Read: i32 = 0;
        pub const Write: i32 = 1;
        pub const Exception: i32 = 2;
    };

};

File: binding/src/libqsocketnotifier.cpp
#include <QEvent>
#include <QObject>
#include <QSocketDescriptor>
#include <QSocketNotifier>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qsocketnotifier.h>
#include "libqsocketnotifier.h"
#include "libqsocketnotifier.hxx"

QSocketNotifier* QSocketNotifier_new(int param1) {
	 return new VirtualQSocketNotifier(static_cast<QSocketNotifier::Type>(param1));
}

QSocketNotifier* QSocketNotifier_new2(intptr_t socket, int param2) {
	 return new VirtualQSocketNotifier(socket, static_cast<QSocketNotifier::Type>(param2));
}

QSocketNotifier* QSocketNotifier_new3(int param1, QObject* parent) {
	 return new VirtualQSocketNotifier(static_cast<QSocketNotifier::Type>(param1), parent);
}

QSocketNotifier* QSocketNotifier_new4(intptr_t socket, int param2, QObject* parent) {
	 return new VirtualQSocketNotifier(socket, static_cast<QSocketNotifier::Type>(param2), parent);
}

libqt_string QSocketNotifier_Tr(const char* s) {
	QString _ret = QSocketNotifier::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSocketNotifier_SetSocket(QSocketNotifier* self, intptr_t socket) {
	self->setSocket(socket);
}

intptr_t QSocketNotifier_Socket(const QSocketNotifier* self) {
	return self->socket();
}

int QSocketNotifier_Type(const QSocketNotifier* self) {
	return self->type();
}

bool QSocketNotifier_IsValid(const QSocketNotifier* self) {
	return self->isValid();
}

bool QSocketNotifier_IsEnabled(const QSocketNotifier* self) {
	return self->isEnabled();
}

void QSocketNotifier_SetEnabled(QSocketNotifier* self, bool enabled) {
	self->setEnabled(enabled);
}

libqt_string QSocketNotifier_Tr2(const char* s, const char* c) {
	QString _ret = QSocketNotifier::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSocketNotifier_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSocketNotifier::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Derived class handler implementation
bool QSocketNotifier_Event(QSocketNotifier* self, QEvent* param1) {
	auto* vqsocketnotifier = dynamic_cast<VirtualQSocketNotifier*>(self);
	if (vqsocketnotifier && vqsocketnotifier->isVirtualQSocketNotifier) {
	return vqsocketnotifier->event(param1);
	} else {
	return self->QSocketNotifier::event(param1);
}
}

// Base class handler implementation
bool QSocketNotifier_QBaseEvent(QSocketNotifier* self, QEvent* param1) {
	auto* vqsocketnotifier = dynamic_cast<VirtualQSocketNotifier*>(self);
	if (vqsocketnotifier && vqsocketnotifier->isVirtualQSocketNotifier) {
vqsocketnotifier->setQSocketNotifier_Event_IsBase(true);
	return vqsocketnotifier->event(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QSocketNotifier_OnEvent(QSocketNotifier* self, intptr_t slot) {
	auto* vqsocketnotifier = dynamic_cast<VirtualQSocketNotifier*>(self);
	if (vqsocketnotifier && vqsocketnotifier->isVirtualQSocketNotifier) {
vqsocketnotifier->setQSocketNotifier_Event_Callback(reinterpret_cast<VirtualQSocketNotifier::QSocketNotifier_Event_Callback>(slot));
}
}

void QSocketNotifier_Connect_activated(QSocketNotifier* self, intptr_t slot) {
    void (*slotFunc)(QSocketNotifier*, QSocketDescriptor*, int) = reinterpret_cast<void (*)(QSocketNotifier*, QSocketDescriptor*, int)>(slot);
    QSocketNotifier::connect(self, &QSocketNotifier::activated, [self, slotFunc](QSocketDescriptor socket, QSocketNotifier::Type activationEvent) {
	slotFunc(self, socket, activationEvent);
    });
}

void QSocketNotifier_Delete(QSocketNotifier* self) {
    delete self;
}

QSocketDescriptor* QSocketDescriptor_new(const QSocketDescriptor* other) {
	 return new QSocketDescriptor(*other);
}

QSocketDescriptor* QSocketDescriptor_new2(QSocketDescriptor* other) {
	 return new QSocketDescriptor(*other);
}

QSocketDescriptor* QSocketDescriptor_new3() {
	 return new QSocketDescriptor();
}

QSocketDescriptor* QSocketDescriptor_new4(intptr_t desc) {
	 return new QSocketDescriptor(desc);
}

QSocketDescriptor* QSocketDescriptor_new5(const QSocketDescriptor* param1) {
	 return new QSocketDescriptor(*param1);
}

QSocketDescriptor* QSocketDescriptor_new6(void** descriptor) {
	 return new QSocketDescriptor(descriptor);
}

void QSocketDescriptor_CopyAssign(QSocketDescriptor* self, QSocketDescriptor* other) {
    *self = *other;
}

void QSocketDescriptor_MoveAssign(QSocketDescriptor* self, QSocketDescriptor* other) {
    *self = std::move(*other);
}

intptr_t QSocketDescriptor_OperatorlongLong(const QSocketDescriptor* self) {
	return self->operator long long();
}

void* QSocketDescriptor_WinHandle(const QSocketDescriptor* self) {
	return self->winHandle();
}

void** QSocketDescriptor_OperatorvoidMultiply(const QSocketDescriptor* self) {
	return self->operator void *();
}

bool QSocketDescriptor_IsValid(const QSocketDescriptor* self) {
	return self->isValid();
}

void QSocketDescriptor_Delete(QSocketDescriptor* self) {
    delete self;
}

